src/operators/batch.rs

```diff
@@ -0,0 +1,92 @@
+use std::{
+    task::Poll,
+    time::{Duration, Instant},
+};
+
+use pyo3::{Python, ToPyObject};
+
+use crate::{pyo3_extensions::TdPyAny, unwrap_any};
+
+use super::stateful_unary::{LogicFate, StatefulLogic};
+
+pub(crate) struct BatchLogic {
+    size: usize,
+    timeout: Duration,
+    last_drain: Instant,
+    acc: Vec<TdPyAny>,
+}
+
+impl BatchLogic {
+    pub(crate) fn builder(size: usize, timeout: Duration) -> impl Fn(Option<TdPyAny>) -> Self {
+        move |resume_snapshot| {
+            let acc = resume_snapshot
+                .and_then(|state| -> Option<Vec<TdPyAny>> {
+                    unwrap_any!(Python::with_gil(|py| state.extract(py)))
+                })
+                .unwrap_or_default();
+            Self {
+                size,
+                acc,
+                timeout,
+                last_drain: Instant::now(),
+            }
+        }
+    }
+
+    /// Drain self.acc, convert it to a TdPyAny and return it.
+    /// If self.acc is empty, return None, but still set self.last_drain.
+    fn drain_acc(&mut self) -> Option<TdPyAny> {
+        self.last_drain = Instant::now();
+        if self.acc.is_empty() {
+            None
+        } else {
+            Some(Python::with_gil(|py| {
+                self.acc
+                    .drain(..)
+                    .collect::<Vec<TdPyAny>>()
+                    .to_object(py)
+                    .into()
+            }))
+        }
+    }
+}
+
+impl StatefulLogic<TdPyAny, TdPyAny, Option<TdPyAny>> for BatchLogic {
+    fn on_awake(&mut self, next_value: Poll<Option<TdPyAny>>) -> Option<TdPyAny> {
+        let timeout_expired = self.last_drain.elapsed() >= self.timeout;
```

Comment:
Mh, your version looks nice, but trying it, to make this work we'd need some changes:
- `eof` has to be initialized before the `if let` on `next_value`, otherwise `next_value`'s content is partially moved and we can't match after that
- we need to check for the size of the batch too, or we only emit at `timeout || eof`
- we also have to check if `self.acc` is empty right before emitting, to avoid emitting empty lists (do we want to? this is probably just a design decision)
- we need to convert `self.acc` to `TdPyAny` to respect the signature

Also, given that the logic is not reused after emitting, we don't really need to update `self.last_drain` right? At this point I'd call that `self.start` and avoid updating it.

So this is what I ended up with:
```rust
let eof = matches!(next_value, Poll::Ready(None));
if let Poll::Ready(Some(value)) = next_value {
    self.acc.push(value);
}

let timeout_expired = self.start.elapsed() >= self.timeout;
let reached_size = self.acc.len() >= self.size;

if (reached_size || timeout_expired || eof) && !self.acc.is_empty() {
    Some(Python::with_gil(|py| {
        std::mem::take(&mut self.acc).to_object(py).into()
    }))
} else {
    None
}
```

What do you think?